### DISCUSSION TITLE
Can I update async atom?

### DISCUSSION THREAD

**Discussion by agumy** [pmndrs/jotai#1910] [Q&A] [answered]
I would like to merge the state obtained by async atom with the result obtained by another async request and update it.
https://codesandbox.io/s/vigorous-allen-r6dicc?file=/src/App.tsx

Currently, this feature is restricted, so atom cannot be updated with this implementation.
I would like to know how to do this for this use case.

https://github.com/pmndrs/jotai/blob/1967bba4f97ba6fdc827d452cf373bb7b66af08f/src/vanilla/store.ts#L423
Will the restrictions be removed in the future?

**Comment by himself65** [comment]
You will need a primitiveAtom that hold the real value

**Reply by agumy** [reply]
I think primitive atoms do not suspend, am I correct?

**Reply by himself65** [reply]
Yes, but in your example, you only need the first render to be suspended so that you can have a primitive with type `Array<T> | null` which null indicates it still pending

**Reply by himself65** [reply]
I've updated the code. https://codesandbox.io/s/rough-forest-stc756?file=/src/App.tsx

**Reply by himself65** [reply]
```tsx
const primiriveAtom = atom<any[] | null>(null);

const fetchUpdate = async () => {
  await new Promise<void>((resolve) => setTimeout(() => resolve(), 2000));
  const response = await fetch("https://jsonplaceholder.typicode.com/todos");
  const json = await response.json();
  return json;
};

const todosAtom = atom(
  async (get) => {
    if (get(primiriveAtom) === null) {
      return fetchUpdate();
    } else {
      return get(primiriveAtom);
    }
  },
  async (get, set, update: any) => {
    const todos = await get(todosAtom);
    set(primiriveAtom, [...todos, update]);
  }
);
```

**Reply by agumy** [reply]
I see. I can certainly make primitive atoms this way. Thank you very much.

**Comment by dai-shi** [comment, marked as answer]
atomWithDefault might be something you are looking for. Alternatively, you could store a promise in a primitive atom.

> Will the restrictions be removed in the future?

Unlikely. It's an API design choice. We should probably update the source code comment.

**Reply by agumy** [reply]
I was able to solve the problem with atomWithDefault, thank you.
